rfcs/side-effect-lifecycle-scheduling.md

```diff
@@ -0,0 +1,124 @@
+                .label(produces(SideEffectLabel))
+                .before(reads(SideEffectLabel))
+                .before(writes(SideEffectLabel))
+                .before(consumes(SideEffectLabel))
+                .add_system(side_effect_producer)
+            )
+        )
+        // TODO: add consumers, readers, and writer system sets
+}
+```
+
+In practice we can't create these system sets like this as it would be very unergonomic for users. But systems sets are just sugar anyways and don't exist at runtime. We can just apply the correct labels to individual systems.
+
+```rs
+// .produces
+.add_system(side_effect_producer.produces(SideEffectLabel))
+// is equivalent to
+.add_system(side_effect_producer
+    .label(SideEffectLabel)
+    .label(produces(SideEffectLabel))
+    .before(reads(SideEffectLabel))
+    .before(writes(SideEffectLabel))
+    .before(comsumes(SideEffectLabel))
+)
+```
+
+> TODO: add desugaring of readers, writers, and consumers.
+
+## Drawbacks
+
+I expect that these relationships will mostly be added automatically when using run criteria, app state, and the like with a higher level api. Users would only be expected to interact with this API when things can't be done automatically or the default behavior doesn't match their needs.
+
+## Open questions
+
+* With looping criteria it might be possible to create a valid dependency cycle that cannot be resolved in the current API. If that is the case then we might need an `.after_can_break` type of edge that could be ignored in dependency cycle checking.
+
+## Alternatives Considered
+
+* The example above for expressing the relationships between side effects using just labeling and system sets has one major problem. There is no current api for adding systems to system sets after the system set has been created. If such an api did exist we could potentially just add systems to the correct system set. This didn't seem better than the proposed API and has some implementation complexity since system sets are mostly just api sugar for labeling multiple systems and does not exist at runtime.
+
+## Future Work
+
+* This is part of a bigger redesign of the ECS trying to solve problems with scheduling around hard sync points and data consistency between states. This RFC by itself doesn't solve those problems, but it does encode information about dependencies that other RFCs might need to actually solve the data consistency problems. 
+* Side effects are effectively typed. Build an API to pass and send side effects and enforce the typing between producers and consumers. This could either be with data stored in the executor or a Resource that manages access through access tokens (what Boxy is working on).
+* How to specify an order to how commands are applied?
+* This API doesn't really directly address data consistency problems due to changes to run criteria. Ideally we'd figure out a way to enforce that the consumer needs to run if side effects have been produced.
```

Comment:
I started to chew on this in my pre-draft RFC here: https://github.com/alice-i-cecile/rfcs/blob/system-graph-flavors/rfcs/34-system-graph-flavors.md#causal-ties-systems-that-must-run-together